File: sorting/intermediate/threeNumberSort.py
def threeNumberSort(array, order):
    firstIdx, secondIdx, thirdIdx = 0, 0, len(array) - 1

    firstValue, secondValue = order[0], order[1]

    while secondIdx <= thirdIdx:
        value = array[secondIdx]

        if value == firstValue:
            array[firstIdx], array[secondIdx] = array[secondIdx], array[firstIdx]
            firstIdx += 1
            secondIdx += 1

        elif value == secondValue:
            secondIdx += 1

        else:
            array[secondIdx], array[thirdIdx] = array[thirdIdx], array[secondIdx]
            thirdIdx -= 1

    return array

File: sorting/intermediate/threeNumberSort.mjs
export default function threeNumberSort(array, ordered) {
    let firstIdx = 0; 
    let secondIdx = 0; 
    let thirdIdx = array.length - 1;
    let firstValue = ordered[0]; 
    let secondValue = ordered[1];

    while (secondIdx <= thirdIdx) {
        const possible = array[secondIdx];

        if (possible === firstValue) {
            swap(firstIdx, secondIdx, array);
            firstIdx++;
            secondIdx++;
        } else if (possible === secondValue) {
            secondIdx++;
        } else {
            swap(secondIdx, thirdIdx, array);
            thirdIdx--;
            secondIdx++;
        }
    }

    return array;
}

function swap(i, j, array) {
    [array[i], array[j]] = [array[j], array[i]];
}

console.log(threeNumberSort([1, 2, 1, 3, 2, 1, 1], [1, 2, 3]));

